fluidsynth.py

```diff
@@ -860,7 +967,54 @@ def get_samples(self, len=1024):
 
         """
         return fluid_synth_write_s16_stereo(self.synth, len)
+    ## Remaining methods in Synth class added by JAllin, July 2021 ff
+       ## Work in progress!!
+    def tuning_dump(self, bank, prog, pitch):
+        return fluid_synth_tuning_dump(self.synth, bank, prog, name.encode(), length(name), pitch)
+
+    def midi_event_get_type(self, event):
+        return fluid_midi_event_get_type(event)
+    def midi_event_get_velocity(self, event):
+        return fluid_midi_event_get_velocity(event)
+    def midi_event_get_key(self, event):
+        return fluid_midi_event_get_key(event)
+    def midi_event_get_channel(self, event):
+        return fluid_midi_event_get_channel(event)
+    def midi_event_get_control(self, event):
+        return fluid_midi_event_get_control(event)
+    def midi_event_get_program(self, event):
+        return fluid_midi_event_get_program(event)
+    def midi_event_get_value(self, event):
+        return fluid_midi_event_get_value(event)
+
+    def play_midi_file(self, filename):
+        self.player = new_fluid_player(self.synth)
+        if self.player == None: return FLUID_FAILED
+        fluid_player_set_playback_callback(self.player, self.custom_router_callback, self.synth)
+        status = fluid_player_add(self.player, filename.encode())
+        if status == FLUID_FAILED: return status
+        playerThread = threading.Thread(target=self.playMidiThread, args=(1,), daemon=True)
+        playerThread.start()
+
+    def playMidiThread(self, name):
+        status = fluid_player_play(self.player)
+        if status == FLUID_FAILED: return status
+        fluid_player_join(self.player)
```

Comment:
You shouldn't need `threading` to run the player in a separate thread - the synth already runs in its own thread. In addition, the player uses the synth's sample timing by default to advance, so calling `fluid_player_join()` shouldn't be needed and might actually cause the player to get out of sync. @drtquin does it not work to just call `fluid_player_play()` in `play_midi_file()`?

Reply:
Hello AZ,

Many thanks for your feedback.

I don't currently have access to my laptop, so can't recheck. However my first attempt was without threading and I found that when playing a MIDI file the rest of my application was locked out.

I'll provide a fuller reply in a couple of weeks when I get my laptop back.

Reply:
Hello AZ,

Finally got my laptop back and a few minutes to spare. You are  of course correct: I've no idea why I seemed to have threading problems.

This means that the player method looks like this:
```
    def play_midi_file(self, filename):
        self.player = new_fluid_player(self.synth)
        if self.player == None: return FLUID_FAILED
        fluid_player_set_playback_callback(self.player, self.custom_router_callback, self.synth)
        status = fluid_player_add(self.player, filename.encode())
        if status == FLUID_FAILED: return status
        status = fluid_player_play(self.player)
        return status
```
How should this change be made?

Reply:
Looks good! If you edit the code in [the branch you made on your own fork](https://github.com/drtquin/pyfluidsynth/tree/drtquin-patch-1) and then commit the changes over there it will update this pull request.

Reply:
Done! Many thanks.